Implement `hash_map` macro

stifskere opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

Implementation of https://github.com/rust-lang/rust/issues/144032

Implements the `hash_map` macro under `std/src/macros.rs`.

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16347320275/job/46184552502) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46184552502) [(plain)](https://github.com/rust-lang/rust/commit/7f00f3486d1ddb4ee1eff5930f3ab762590540f4/checks/46184552502/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16348888417/job/46190167432) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46190167432) [(plain)](https://github.com/rust-lang/rust/commit/3720f96d7b88f714f9745e586e23b5437c5e6daf/checks/46190167432/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

NyxCode commented:
Wouldn't using one of `HashMap::with_capacity()`, `reserve()`, `FromIterator<(K, V)>`, `Extend<(K, V)>` or `From<[(K, V); N]>` be better, avoiding unnecessary allocations?

stifskere commented:
> Wouldn't using one of `HashMap::with_capacity()`, `HashMap::reserve()`, `FromIterator<(K, V)>`, `Extend<(K, V)>` or `From<[(K, V); N]>` be better, avoiding unnecessary allocations?

I was thinking about `HashMap::with_capacity()` but I couldn't see how I could convert a repetition into a number.

`From...` would essentially be the same, just reallocation, so for clarity I just expanded into `insert`.

I could see that `0 $(+ 1)*` could be evaluated constantly by the compiler, thus we could get a size, but I'm not sure if it's the best way.

I will add a question about the size in the tracking issue.

Noratrieb commented on library/std/src/macros.rs:
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
+/// [`HashMap::new`]: crate::collections::HashMap::new
+/// [`HashMap::insert`]: crate::collections::HashMap::insert
+#[macro_export]
+#[allow_internal_unstable(hint_must_use)]
```
you don'takw use of this feature (maybe it would make sense to?)

tmke8 replied:
(Noratrieb's point was that you actually have to use `must_use` within the macro. This line only makes it so that using `must_use` within the macro is *allowed*.)

Noratrieb commented on library/std/src/macros.rs:
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
+    }};
+
+    ( $( $key:expr => $value:expr ),* $(,)? ) => {{
+        let mut __map = ::std::collections::HashMap::new();
```
you can just call it `map`, no underscores needed thanks to hygiene

stifskere commented:
> All the options i mentioned should be equivalent to `with_capacity` as far as I can tell, though I don't know if the optimizer can look through all options equally well (e.g. wrt. stack usage). Both these should work, for example:
> 
> ```rust
> let len = 1 $(+ {stringify!($k); 1})*;
> let mut map = HashMap::with_capacity(len);
> $( map.insert($k, $v); )*
> ```
> 
> ```rust
> HashMap::from([$(($k, $v)),*])
> ```

I think I'd rather go for intrinsics, except since `HashMap` is a wrapper, I don't think there are many. I feel that if someone decides to expand the macro via IDE or cargo expand, they will find a lot of strings and be like "???". Probably open an issue later.

stifskere commented:
> I don't think intrinsics make sense here, this is a library type. I think the repeat insert method makes the most sense as it doesn't have the potential stack usage problem of from. Repeat insert is also how extend is implemented under the hood: https://github.com/rust-lang/hashbrown/blob/670213fb32d208759d0331996894e05604de0c18/src/map.rs#L4562.
> 
> with_capacity definitely makes sense though. You can count the elements like this: https://lukaswirth.dev/tlborm/decl-macros/building-blocks/counting.html

I would go for the [slice length](https://lukaswirth.dev/tlborm/decl-macros/building-blocks/counting.html#slice-length) approach, but resolved constantly, which is technically what I'm doing just that instead of making a variable using `replace_expr!`

Noratrieb commented:
> Oh actually you might be able to use the unstable macro_metavar_expr_count for the counting too, I forgot about that :D

You should try this, which I only remembered after posting the first message

stifskere commented:
> > Oh actually you might be able to use the unstable macro_metavar_expr_count for the counting too, I forgot about that :D
> 
> You should try this, which I only remembered after posting the first message 

The problem with this is that it would delay stabilization a lot, so I'm not sure, I don't think adding unstable to unstable helps

tmke8 commented:
Would it be possible to use `HashMap::default()` instead of `HashMap::new()` so that this macro can be used with non-default hashing algorithms as well? (Though this might cause type inference problems...)

stifskere commented:
> Would it be possible to use `HashMap::default()` instead of `HashMap::new()` so that this macro can be used with non-default hashing algorithms as well? (Though this might cause type inference problems...)
> 
> EDIT: or rather
> 
> ```rust
> HashMap::with_capacity_and_hasher(capacity, Default::default())
> ```
> 
> for specifying capacity.

We are discussing to use `with_capacity` instead, If you could give your opinion on getting capacity from a macro repetition, possibly would be better, because knowing the size before creation avoids reallocation and makes the thing faster overall.

stifskere commented:
I personally don't find a constant evaluation that consuming, nor for the compiler, it's just allocating an array of sized 0 elements with 1 byte alignment and then counting it.

slanterns commented:
https://rust-lang.github.io/rfcs/3086-macro-metavar-expr.html
https://github.com/rust-lang/rust/tree/master/tests/ui/macros/metavar-expressions

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16653216349/job/47131389281) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47131389281) [(plain)](https://github.com/rust-lang/rust/commit/1a6d862c0b274928114ad198cea509d1b9a2d306/checks/47131389281/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

tmke8 commented:
> I'm not sure if I could think of a syntax for that to be added, you could comment your proposal on the tracking issue.

I didn't mean to change the syntax, just the implementation:

```rust
macro_rules! hash_map {

    () => {{
        ::std::collections::HashMap::new()
    }};

    ( $( $key:expr => $value:expr ),* $(,)? ) => {{
        let mut map = ::std::collections::HashMap::with_capacity_and_hasher(
            ${count($key)} , Default::default()
        );
        $( map.insert($key, $value); )*
        map
    }}
}

let map: FxHashMap = hash_map!["foo".to_string() => 0];
```

Though as I said I'm not sure about type inference. But you're right I should post in the tracking issue instead.

Noratrieb commented on library/std/src/macros.rs:
````diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
+/// This macro is used by concatenating comma separated sequences
+/// of `<key:expr> => <value:expr>`.
+///
+/// An example usage of this macro could be the following
+///
+/// ```rust
````
```suggestion
/// # Examples
///
/// ```
```

Noratrieb commented on library/std/src/macros.rs:
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
+///     "key1" => "value1"
+/// };
+///
+/// for (key, value) in map {
+///     println!("{key:?} => {value:?}");
+/// }
```
can you replace this with assertions instead? to make it clearer what the result is like (though it is pretty obvious) and also serve as a test.

Noratrieb commented on library/std/src/macros.rs:
````diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
+/// ```rust
+/// #![feature(hash_map_macro)]
+///
+/// let map = hash_map!{
````
```suggestion
/// let map = hash_map! {
```
rustfmt puts a space before braced macro invocations.

Noratrieb commented on library/std/src/macros.rs:
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
+/// Creates a [`HashMap`] containing the arguments.
+///
+/// This macro creates or either an empty [`HashMap`] using
+/// [`HashMap::new`] if there are no arguments, or generates
+/// a [`HashMap::insert`] call for each element.
```
I'm not sure it's useful to be this precise about `insert`, I would probably avoid mentioning it. Mentioning it this clearly may imply that it's a guarantee (not sure how someone would rely on this though) which it is not. Anyone curious to see the implementation can view the source to see the `insert`.

stifskere replied:
Should I mention the move semantics in some other way, or simply skip this completely?

Noratrieb replied:
This is my attempt at the documentation, I rephrased a few things to attempt to make it a bit clearer and focus on the more important information:

```rust
/// Creates a [`HashMap`] containing the arguments.
/// 
/// `hash_map!` allows specifying the entries that make
/// up the [`HashMap`] where the key and value are separated by a `=>`.
/// 
/// The entries are separated by commas, where a trailing comma can be used.
/// 
/// It is semantically equivalent to using repeated [`HashMap::insert`]
/// on a newly created hashmap.
/// `hash_map!` will attempt to avoid repeated reallocations by using [`with_capacity`].
///
/// # Examples
///
/// ```rust
/// #![feature(hash_map_macro)]
///
/// let map = hash_map! {
///     "key" => "value",
///     "key1" => "value1"
/// };
///
/// assert_eq!(map.get("key"), Some(&"value"));
/// assert_eq!(map.get("key1"), Some(&"value1"));
/// ```
///
/// ```rust
/// #![feature(hash_map_macro)]
///
/// let map = hash_map! {
///     "key" => "value", // notice the ,
/// };
///
/// assert_eq!(map.get("key"), Some(&"value"));
/// assert_eq!(map.get("no-key"), None);
/// ```
///
/// [`HashMap`]: crate::collections::HashMap
/// [`HashMap::insert`]: crate::collections::HashMap::insert
/// [`with_capacity`]: crate::collections::HashMap::with_capacity
```

You can adapt some stuff from there (or take the whole thing if you think it's good). We don't necessarily have to make the documentation perfect in this PR, but we can try a bit to make it nice

Noratrieb replied:
This is my attempt at the documentation, I rephrased a few things to attempt to make it a bit clearer and focus on the more important information:

```rust
/// Creates a [`HashMap`] containing the arguments.
/// 
/// `hash_map!` allows specifying the entries that make
/// up the [`HashMap`] where the key and value are separated by a `=>`.
/// 
/// The entries are separated by commas, where a trailing comma can be used.
/// 
/// It is semantically equivalent to using repeated [`HashMap::insert`]
/// on a newly created hashmap.
/// `hash_map!` will attempt to avoid repeated reallocations by using [`with_capacity`].
///
/// # Examples
///
/// ```rust
/// #![feature(hash_map_macro)]
///
/// let map = hash_map! {
///     "key" => "value",
///     "key1" => "value1"
/// };
///
/// assert_eq!(map.get("key"), Some(&"value"));
/// assert_eq!(map.get("key1"), Some(&"value1"));
/// ```
///
/// ```rust
/// #![feature(hash_map_macro)]
///
/// let map = hash_map! {
///     "key" => "value", // notice the ,
/// };
///
/// assert_eq!(map.get("key"), Some(&"value"));
/// assert_eq!(map.get("no-key"), None);
/// ```
///
/// [`HashMap`]: crate::collections::HashMap
/// [`HashMap::insert`]: crate::collections::HashMap::insert
/// [`with_capacity`]: crate::collections::HashMap::with_capacity
```

You can adapt some stuff from there (or take the whole thing if you think it's good). We don't necessarily have to make the documentation perfect in this PR, but we can try a bit to make it nice

Noratrieb commented on library/std/src/macros.rs:
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
+    }};
+
+    ( $( $key:expr => $value:expr ),* $(,)? ) => {{
+        let mut map = ::std::collections::HashMap::with_capacity( ${count($key)} );
```
```suggestion
        let mut map = $crate::collections::HashMap::with_capacity( ${count($key)} );
```

Noratrieb commented on library/std/src/macros.rs:
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
+#[unstable(feature = "hash_map_macro", issue = "144032")]
+macro_rules! hash_map {
+    () => {{
+        ::std::collections::HashMap::new()
```
```suggestion
        $crate::collections::HashMap::new()
```

Noratrieb commented on library/std/src/macros.rs:
````diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
+/// }
+/// ```
+///
+/// Note that since this macro only generates [`HashMap::insert`]
+/// calls, the move semantics for the values are the same as
+/// the insert method.
````
I'm also not sure whether this sentence adds useful information.
Maybe you can add a sentence like this at the end of the top paragraph:

> The key and value are moved into the hashmap.

although I would consider the ownership semantics here fairly obvious, but a short sentence like the one above probably wouldn't hurt.

Noratrieb reviewed (commented):
Looks like you figured out the `macro_metavar_expr`, that's great. It looks a lot better with `count` :). I see that it doesn't quite work though yet, I'm not sure why exactly.. it might be that the feature gating is not handled in the compiler in a way that makes `allow_internal_unstable` work..

I took a closer look at the implementation and documentation and wrote down a few minor notes. Additionally, can you add a few tests for example with trailing commas or an empty invocation?

Overall it looks good, so after you address the feedback and add the tests we can merge it (please do squash the commits into one before that though)!

Noratrieb commented on library/std/src/macros.rs:
```diff
@@ -379,3 +379,51 @@ macro_rules! dbg {
+/// [`HashMap::new`]: crate::collections::HashMap::new
+/// [`HashMap::insert`]: crate::collections::HashMap::insert
+#[macro_export]
+#[allow_internal_unstable(macro_metavar_expr)]
```
Yeah it looks like I can't make easily make this work. I asked on Zulip if anyone else does: https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/allow_internal_unstable.28macro_metavar_expr.29

stifskere replied:
Right, forget the anterior comment, I misread, should I revert this to static evaluation or a simple `new`?

Noratrieb replied:
Adding new public macros would need another ACP and shouldn't be done here, but it seems unlikely that such a macro would be accepted. Just create an unstable doc(hidden) helper, which you can make a bit more generic if you want to (but I probably wouldn't do that)

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16662934715/job/47163855031) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47163855031) [(plain)](https://github.com/rust-lang/rust/commit/c19f87ad3c71b01c354e26738902e54b0ff00eb9/checks/47163855031/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

stifskere commented:
I yesterday was running solely with a coffee lol.

stifskere commented:
I was yesterday solely running on a coffee, I now realize how bad that was.

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16667122346/job/47175606036) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47175606036) [(plain)](https://github.com/rust-lang/rust/commit/79d143f426c6b3ff3f3ba586c89f1dc68cefbd73/checks/47175606036/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

stifskere commented:
I think a macro is not a good idea, we semingly end with the same feature gate problem, I'l make some sort of hack with a const function instead, but without lsp I need to investigate where I can put that

Noratrieb commented:
In this case you're just missing another `allow_internal_unstable` on your `repetition_utils`, since `allow_internal_unstable` isn't transitive:
```patch
diff --git a/library/std/src/macros.rs b/library/std/src/macros.rs
index 7055168e0a6..455d0ca9e24 100644
--- a/library/std/src/macros.rs
+++ b/library/std/src/macros.rs
@@ -383,6 +383,7 @@ macro_rules! dbg {
 #[doc(hidden)]
 #[macro_export]
 #[unstable(feature = "hash_map_internals", issue = "none")]
+#[allow_internal_unstable(hash_map_internals)]
 macro_rules! repetition_utils {
     (@count $($tokens:tt),*) => {{
         [$($crate::repetition_utils!(@replace $tokens => ())),*].len()
```
This makes it work for me.

stifskere commented:
> In this case you're just missing another `allow_internal_unstable` on your `repetition_utils`, since `allow_internal_unstable` isn't transitive:
> 
> ```diff
> diff --git a/library/std/src/macros.rs b/library/std/src/macros.rs
> index 7055168e0a6..455d0ca9e24 100644
> --- a/library/std/src/macros.rs
> +++ b/library/std/src/macros.rs
> @@ -383,6 +383,7 @@ macro_rules! dbg {
>  #[doc(hidden)]
>  #[macro_export]
>  #[unstable(feature = "hash_map_internals", issue = "none")]
> +#[allow_internal_unstable(hash_map_internals)]
>  macro_rules! repetition_utils {
>      (@count $($tokens:tt),*) => {{
>          [$($crate::repetition_utils!(@replace $tokens => ())),*].len()
> ```
> 
> This makes it work for me.

Done, I added it, let's see if tests run correctly now.

By the way, what's the way you test whether tests fail and whether stability is correct locally? Because I would like to do the same instead of relying on running the workflows. I have read some about the `./x` bootstrap, I need to figure out how to configure it locally.

Noratrieb commented:
I can recommend reading this to see how to build and run the compiler with a standard library: https://rustc-dev-guide.rust-lang.org/building/quickstart.html
I did `./x build --stage 1 library` to build the compiler and have it linked up as mentioned in the guide to then be able to do `rustc +dev mytest.rs` to test building the file (where I invoke the macro).

Noratrieb commented:
(somehow github doesn't show it nicely despite me leaving it as a review comment but i added a comment here in case it got lost in the ui: https://github.com/rust-lang/rust/pull/144070#discussion_r2248866788)

stifskere commented:
> (somehow github doesn't show it nicely despite me leaving it as a review comment but i added a comment here in case it got lost in the ui: [#144070 (comment)](https://github.com/rust-lang/rust/pull/144070#discussion_r2248866788))

I believe it didn't show up because it was solved, I didn't see the comment at all, If you could re-review the actual documentation, I changed it a bit.

stifskere committed:
feat: implement `hash_map!` macro

bors commented:
:pushpin: Commit 066023e47c0c92e7edefd60831ce7d6f15f23750 has been approved by `Noratrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Noratrieb 066023e47c0c92e7edefd60831ce7d6f15f23750 -->
<!-- homu: {"type":"Approved","sha":"066023e47c0c92e7edefd60831ce7d6f15f23750","approver":"Noratrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

danielhenrymantilla commented on library/std/src/macros.rs:
```diff
@@ -379,3 +379,77 @@ macro_rules! dbg {
         ($($crate::dbg!($val)),+,)
     };
 }
+
+#[doc(hidden)]
+#[macro_export]
+#[allow_internal_unstable(hash_map_internals)]
+#[unstable(feature = "hash_map_internals", issue = "none")]
+macro_rules! repetition_utils {
+    (@count $($tokens:tt),*) => {{
+        [$($crate::repetition_utils!(@replace $tokens => ())),*].len()
+    }};
+
+    (@replace $x:tt => $y:tt) => { $y }
+}
```
In which case this would not be necessary:

```suggestion
```

danielhenrymantilla commented on library/std/src/macros.rs:
```diff
@@ -379,3 +379,77 @@ macro_rules! dbg {
+        let mut map = $crate::collections::HashMap::with_capacity(
+            const { $crate::repetition_utils!(@count $($key),*) }
+        );
+        $( map.insert($key, $value); )*
```
```suggestion
        $( map.insert($key, $value); )+
```

danielhenrymantilla reviewed (commented):
Looks good!

🤔 I wonder how the implementation would compare to doing:

```rs
<
    $crate::collections::HashMap<_, _, $crate::hash::RandomState>
    as
    $crate::convert::From<[(_, _); _]>
>::from([
    $(
        ($key, $value)
    ),+ // empty case still handled as `new()`.
])
```

Intuitively, doing a batch of `.insert()`s does require the `map` to go back to a fully usable-from-the-outside state, whereas a direct batched constructor such as `From<[(K, V); N]>` does not, so depending on the impl, there might be more room for optimisation with this more direct approach 🤔

Noratrieb commented on library/std/src/macros.rs:
```diff
@@ -379,3 +379,77 @@ macro_rules! dbg {
+/// [`HashMap::insert`]: crate::collections::HashMap::insert
+/// [`HashMap::with_capacity`]: crate::collections::HashMap::with_capacity
+#[macro_export]
+#[allow_internal_unstable(hash_map_internals)]
```
doesn't work :)

Noratrieb commented:
I have checked the `FromIterator` implementations, they are not any more efficient than repeat inserts (but the array impl will blow the stack :D)

danielhenrymantilla commented:
Thanks for looking into these 🙏

stifskere commented:
> I have checked the `FromIterator` implementations, they are not any more efficient than repeat inserts (but the array impl will blow the stack :D)

I was now checking about this you said, which feels weird, I thought that without recursiveness this wouldn't have a problem, so should I add another PR to the tracking issue? Wait for a comment period?
